zio-http/src/main/scala/zhttp/service/Handler.scala

```diff
@@ -106,87 +107,84 @@ private[zhttp] final case class Handler[R](
-              releaseRequest(jReq)
-
-            case HttpData.BinaryByteBuf(data) =>
-              unsafeWriteLastContent(data)
-              releaseRequest(jReq)
-
-            case data @ HttpData.BinaryChunk(_) =>
-              unsafeWriteLastContent(data.encodeAndCache(res.attribute.memoize))
-              releaseRequest(jReq)
-
-            case HttpData.BinaryStream(stream) =>
-              unsafeRunZIO(
-                writeStreamContent(stream) *> Task(releaseRequest(jReq)),
-              )
-          }
-        }
+        processResponse(jReq, res)
+      case HExit.Failure(e)   =>
+        processErrorResponse(jReq, Some(e))
+      case HExit.Empty        =>
+        processErrorResponse(jReq, None)
+    }
+
+  }
+
+  private def processResponse(jReq: FullHttpRequest, res: Response[R, Throwable])(implicit
+    ctx: ChannelHandlerContext,
+  ) = {
+    if (self.isWebSocket(res)) {
+      self.upgradeToWebSocket(ctx, jReq, res)
+    } else {
+      val responseWithHeadersAndByteBuf = attachHeadersWithComputedBuf(res)
+      unsafeWriteAnyResponse(responseWithHeadersAndByteBuf)
+
+      responseWithHeadersAndByteBuf.data match {
+        case HttpData.Empty =>
+          unsafeWriteAndFlushLastEmptyContent()
+
+        case data @ HttpData.Text(_, _) =>
+          unsafeWriteLastContent(data.encodeAndCache(res.attribute.memoize))
+
+        case HttpData.Complete(data) =>
+          unsafeWriteLastContent(data)
 
-      case HExit.Failure(e) =>
-        unsafeWriteAndFlushErrorResponse(e)
-        releaseRequest(jReq)
-      case HExit.Empty      =>
-        unsafeWriteAndFlushEmptyResponse()
-        releaseRequest(jReq)
+        case data @ HttpData.BinaryChunk(_) =>
+          unsafeWriteLastContent(data.encodeAndCache(res.attribute.memoize))
+
+        case HttpData.BinaryStream(stream) =>
+          unsafeRunZIO(
+            writeStreamContent(stream),
+          )
+      }
+      releaseRequest(jReq)
+    }
+  }
+
+  private def processErrorResponse(jReq: FullHttpRequest, err: Option[Throwable])(implicit
+    ctx: ChannelHandlerContext,
+  ) = {
+    err.fold(unsafeWriteAndFlushEmptyResponse())(unsafeWriteAndFlushErrorResponse)
+    releaseRequest(jReq)
+  }
+
+  private def attachHeadersWithComputedBuf(res: Response[R, Throwable]): Response[R, Throwable] = {
+    val modifiedData = res.data match {
+      case data @ HttpData.Text(_, _)     => HttpData.fromByteBuf(data.encodeAndCache(res.attribute.memoize))
+      case data @ HttpData.BinaryChunk(_) => HttpData.fromByteBuf(data.encodeAndCache(res.attribute.memoize))
+      case _                              => res.data
     }
+    val size         = modifiedData.unsafeSize
+    val isChunked    = modifiedData.isChunked
+    Response(
+      res.status,
+      res.headers ++
+        Headers(Name.ContentLength -> size.toString).when(size >= 0) ++
+        Headers(Name.TransferEncoding -> Value.Chunked).when(isChunked),
+      modifiedData,
+      res.attribute,
+    )
```

Comment:
Again a lot more allocations happening here unnecessarily. I think this too will affect the performance of the system. I would recommend in this PR keep the implementation exactly as it was, except the content-length problem. 

Also I observed, there aren't enough tests asserting for `content-length` .

Reply:
Sure, will add asserts as well.